WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+  flow from creation to rejection. E.g, for the
+  [`loaded`](https://drafts.csswg.org/css-font-loading-3/#dom-fontface-loaded)
+  property of a [`FontFace`](https://drafts.csswg.org/css-font-loading-3/#fontface)
+  instance, creating the `FontFace` instance causes both the promise creation
+  and the loading of the font.
+- More general state transitions are similar to one-time “events” in that each
+  promise is created at the same time as an async operation starts that will
+  resolve it.
+
+## Cross-document navigations
+
+When a cross-document navigation happens, even if it is same-origin, the context
+will be reset such that document load and tasks that directly flow from it
+(including execution of classic scripts found during parsing) run with the
+initial AsyncContext snapshot. The initial AsyncContext Snapshot will be
+provided by the host environment (i.e. by the HTML spec, for the web), and it
+should either be empty, or only have mappings from spec-defined
+`AsyncContext.Variable`s (see “Using AsyncContext from web specs” below).
+
+## Module evaluation
+
+When you import a JS module multiple times, it will only be fetched and
+evaluated once. Since module evaluation should not be racy (i.e. it should not
+depend on the order of various imports), the context should be reset so that
+module evaluation always runs with the initial AsyncContext snapshot. Inside of
+`ShadowRealm`s, the AsyncContext snapshot for all module evaluations will be the
+snapshot which was active when the `ShadowRealm` was created.
+
+# Editorial aspects of AsyncContext integration in web specifications
+
+An agent always has an associated AsyncContext mapping, in its
+`[[AsyncContextMapping]]` field. When the agent is created, this mapping will be
+set to an HTML-provided initial state, but JS user code can change it in a
+strictly scoped way.
+
+The only way JS code can modify the current mapping is through
+`AsyncContext.Variable` and `AsyncContext.Snapshot`’s `run()` methods, which
+switch the context before calling a callback and switch it back after it
+synchronously returns or throws. This ensures that for purely synchronous
+execution, the context is automatically propagated along the data flow. It is
+when tasks and microtasks are queued that web specs must manually track the data
+flow.
+
+The TC39 proposal spec text includes two abstract operations that web specs can
+use to store and switch the context:
+- [`AsyncContextSnapshot()`](https://tc39.es/proposal-async-context/#sec-asynccontextsnapshot)
+  returns the current AsyncContext mapping.
+- [`AsyncContextSwap(context)`](https://tc39.es/proposal-async-context/#sec-asynccontextswap)
+  sets the current AsyncContext mapping to `context`, and returns the previous
+  one. `context` must only be a value returned by one of these two operations.
+
+We propose adding a web spec algorithm “run the AsyncContext Snapshot”, that could be used like this:
```

Comment:
This seems like it will be annoying for web spec authors to remember to use.

When would they need to use this, given that Web IDL promise and callback integration should take care of almost everything? E.g. in which of the above examples?